gunrock/app/pr_nibble/pr_nibble_enactor.cuh

```diff
@@ -0,0 +1,539 @@
+            
+            bool already_touched = atomicMax(touched + dest, 1) == 1;
+            return !already_touched;
+        };
+
+        // filter operation
+        auto filter_op = [
+        ] __host__ __device__ (
+            const VertexT &src, VertexT &dest, const SizeT &edge_id,
+            const VertexT &input_item, const SizeT &input_pos,
+            SizeT &output_pos) -> bool
+        {
+            return true;
+        };
+        
+        // --
+        // Run
+        
+        GUARD_CU(frontier.V_Q()->ForAll(
+            compute_op,
+            frontier.queue_length,
+            util::DEVICE,
+            oprtr_parameters.stream
+        ));
+
+        // GUARD_CU2(cudaDeviceSynchronize(),
+        //     "cudaDeviceSynchronize failed");
+
+        GUARD_CU(oprtr::Advance<oprtr::OprtrType_V2V>(
+            graph.csr(), frontier.V_Q(), frontier.Next_V_Q(),
+            oprtr_parameters, advance_op, filter_op));
+
+        // GUARD_CU2(cudaDeviceSynchronize(),
+        //     "cudaDeviceSynchronize failed");
+        
+        if (oprtr_parameters.advance_mode != "LB_CULL" &&
+            oprtr_parameters.advance_mode != "LB_LIGHT_CULL")
+        {
+            frontier.queue_reset = false;
+            GUARD_CU(oprtr::Filter<oprtr::OprtrType_V2V>(
+                graph.csr(), frontier.V_Q(), frontier.Next_V_Q(),
+                oprtr_parameters, filter_op));
+        }
+
+        // Get back the resulted frontier length
+        GUARD_CU(frontier.work_progress.GetQueueLength(
+            frontier.queue_index, frontier.queue_length,
+            false, oprtr_parameters.stream, true));
+        
+        // Convergence checking
+        ValueT grad_thresh = rho * alpha * (1 + eps);
+
+        auto &d_grad_scale = data_slice.d_grad_scale;
+        GUARD_CU(cudaMemset(d_grad_scale, 0, 1 * sizeof(int)));
+        
+        auto convergence_op = [
+            graph,
+            grad,
+            d_grad_scale,
+            grad_thresh,
+            iteration,
+            src_node,
+            alpha,
+            num_ref_nodes
+        ] __host__ __device__ (VertexT &v) {
+            
+            ValueT v_dn_sqrt  = 1.0 / sqrt((ValueT)graph.GetNeighborListLength(v));
+            
+            ValueT val = abs(grad[v] * v_dn_sqrt);
+            
+            if(v == src_node) 
+                val -= (alpha / num_ref_nodes) * v_dn_sqrt;
+            
+            if(val > grad_thresh)
+                atomicMax(d_grad_scale, 1);
+        };
+        
+        GUARD_CU(frontier.V_Q()->ForEach(
+            convergence_op,
+            frontier.queue_length,
+            util::DEVICE,
+            oprtr_parameters.stream
+        ));
+        
+        cudaDeviceSynchronize();
+        cudaMemcpy(data_slice.h_grad_scale, d_grad_scale, 1 * sizeof(int), cudaMemcpyDeviceToHost);
+        
+        GUARD_CU2(cudaDeviceSynchronize(),
+            "cudaDeviceSynchronize failed");
+        
+        return retval;
+    }
```

Comment:
I am trying to think if the operators you have here can be somehow reduced a bit, currently there is kernel fusion for advance and filter, but operators are serial execution (with I believe a device sync, therefore costly). I don't think I can provide much input on the implementation (and everything else looks pretty straight forward), but it is worth investing some time to see what we can do better with ops.

Reply:
Can you explain what you mean?  I'm not sure I understand

Reply:
`compute_op` -> `ForAll`
`advance_op` and `filter_op`
convergence_op -> `ForEach`

These three steps are in serial, but within them there should be a lot of parallelism.
ForAll and ForEach though are not load balanced.

It checks for convergence after the advance + filter, so I was wondering if there is a way to represent this algorithm in a way where we can utilize advance and filter more or reduce the loops (because they have a global barrier afterwards, and barriers are costly in CUDA).

Side note: the implementation itself worked perfectly, I was just noting down that there may be room for improvement in terms of performance.

Reply:
Yeah interesting.

I don't think we can check for convergence in the exact same way as the original implementation w/o a barrier between `advance_op` and `convergence_op` -- `convergence_op` has to wait for all `grad` to be computed. 

However, we could merge the `compute_op` and the `advance_op` by doing something like:
```
bool src_was_touched = atomicOr(src_touched[src], true);
if(!src_was_touched) compute_op(src);
```
at the top of `advance_op`.  Do you think that might help w/ performance?

Reply:
Yes, I think so. You now don't have a global barrier, also utilizing load balancing strategies of advance and you have good amount of work in the whole kernel.

Reply:
OK cool, I can make that change.

Can you confirm that's a reasonable way to "mark" a node?

Reply:
Are multiple nodes trying to mark the node valid?

Reply:
-- I am thinking if all of the nodes are just trying to mark the active node valid, then do you really need to have an atomic there? (another way of asking this question is are there any nodes marking this `false`)?

Reply:
Hm this gets a little more complicated actually:  all the other threads w/ the same `src` value have to block until `compute_op(src)` is done.  Is there a way to do that without a global barrier?

Reply:
Yes, you use `__syncthreads()` for this, where you want all threads to be done with whatever `compute_op` needs to do.

Reply:
__syncthreads() is cheaper than doing a global sync, but notice that this is a barrier per block.

Reply:
This may be assuming that the `src` you are working on (all threads working on that) are in a single block. Which could be possible, gunrock a lot of times has a filling launch with fewer blocks (e.g. 2) and lots of threads (e.g. 1024).